### DISCUSSION TITLE
use of external LM  to improve transcriptions

### DISCUSSION THREAD

**Discussion by iggygeek** [k2-fsa/icefall#940] [Q&A] [answered]
Could you explain how to used an external LM, N-gram or RNN to improve the trnascriptions ?

**Comment by marcoyang1998** [comment, marked as answer]
We provide several decoding methods with external language models. Please have a look at https://github.com/k2-fsa/icefall/blob/master/egs/librispeech/ASR/RESULTS.md#pruned_transducer_stateless7-zipformer.

**Comment by iggygeek** [comment]
OK, but why do you use only a BPE-based N-gram LM instead of a word-level N-gram with a lexicon.txt of the type word -> BPE tokens ?

**Comment by marcoyang1998** [comment]
We support word level N-gram as well, please have a look at here https://github.com/k2-fsa/icefall/blob/9ddd811925534dc47b183a23429a4727c6416e81/egs/librispeech/ASR/pruned_transducer_stateless2/beam_search.py#L101

**Comment by iggygeek** [comment]
Well I tried that a word-level ARPA LM with  a lexicon.txt of the type word -> BPE tokens  using the BPE models used for training.
I validated using validate_lang_bpe.py, but I got empty hypotheses when decoding with fast_beam_search_nbest_LG ...

**Comment by marcoyang1998** [comment]
Could you please share the decoding command you are using?

BTW, using neural network LMs usually gives a larger performance improvement. If you want to have optimal performance, we recommend you use [`modified_beam_search_LODR`](https://github.com/k2-fsa/icefall/blob/cad6735e0739f149ba3f452e52a948da946527dc/egs/librispeech/ASR/pruned_transducer_stateless2/beam_search.py#L1851).

**Comment by iggygeek** [comment]
I think the problem was that the script generating LG was using a default path for the ARPA LM, whereas I thought that it would looked for it in the lang_bpe_XXX directory...
